app/recipe/tests/test_recipe_api.py

```diff
@@ -0,0 +1,210 @@
+                                     key=lambda ingredient: ingredient['name'])
+        recipe_ingredients = sorted(res.data['ingredients'],
+                                    key=lambda ingredient: ingredient['name'])
+        self.assertEqual(payload_ingredients, recipe_ingredients)
+        self.assertEqual(res.data['name'], payload['name'])
+        self.assertEqual(res.data['description'], payload['description'])
+
+    def test_create_recipe_with_invalid_ingredients(self):
+        """Test creating a recipe with blank ingredient name"""
+        payload = {
+            'name': 'Cheeseburger',
+            'description': 'Yummy',
+            'ingredients': [
+                {'name': 'Meat'},
+                {'name': 'Bun'},
+                {'name': ''}
+            ]
+        }
+
+        res = self.client.post(RECIPES_URL, payload, format='json')
+        self.assertEqual(res.status_code, status.HTTP_400_BAD_REQUEST)
+
+    def test_create_recipe_invalid_name(self):
+        """Test creating recipe without name"""
+        payload = {'name': ''}
+        res = self.client.post(RECIPES_URL, payload)
+
+        self.assertEqual(res.status_code, status.HTTP_400_BAD_REQUEST)
+
+    def test_retrieve_recipes(self):
+        """Test retrieving recipe api data"""
+        sample_recipe(name='Cheeseburger')
+        sample_recipe(name='Soup')
+
+        res = self.client.get(RECIPES_URL)
+        recipes = Recipe.objects.all()
+        serializer = RecipeSerializer(recipes, many=True)
+
+        self.assertEqual(res.status_code, status.HTTP_200_OK)
+        self.assertEqual(res.data, serializer.data)
+
+    def test_update_recipe_new_ingredients_successful(self):
+        """Test updating recipe with list of ingredients"""
+        recipe = sample_recipe(name='Cheeseburger', description='Yummy')
+        payload = {
+            'id': recipe.id,
+            'name': recipe.name,
+            'description': recipe.description,
+            'ingredients': [
+                {'name': 'Cheese'}, {'name': 'Bun'}, {'name': 'Meat'}
+            ]
+        }
+        url = detail_url(recipe.id)
+        res = self.client.put(url, payload, format='json')
+        recipe.refresh_from_db()
+        serializer = RecipeSerializer(recipe)
+
+        self.assertEqual(res.status_code, status.HTTP_200_OK)
+        self.assertEqual(serializer.data, res.data)
+
+    def test_update_recipe_all_fields_successful(self):
+        """Test updating recipe with replacement ingredients"""
+
+        """Create initial recipe with ingredients first"""
+        recipe = sample_recipe(name='Cheeseburger', description='Yummy')
+        payload = {
+            'id': recipe.id,
+            'name': recipe.name,
+            'description': recipe.description,
+            'ingredients': [
+                {'name': 'Cheese'}, {'name': 'Bun'}, {'name': 'Meat'}
+            ]
+        }
+        url = detail_url(recipe.id)
+        self.client.put(url, payload, format='json')
```

Comment:
why do you put this initial `payload` if later the tests are about the `payload2`?

Reply:
Agreed, not necessary and should be removed.

Reply:
Oh, actually my thinking here was to verify that the "old" ingredients no longer appear in the ingredients list after I put with payload2.